org-timeline.el

```diff
@@ -146,27 +142,56 @@ Return new copy of STRING."
-                (push (list beg end face) tasks)))))))
-    (setq tasks (nreverse tasks))
+              (push (list beg end face) tasks))))))
+    (nreverse tasks)))
+
+(defun org-timeline--generate-hourline ()
+  "Generate the template for the timeline's hourline."
+  (let* ((hourline "|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|  :00|")
+         (start-hour (abs org-timeline-start-hour))
+         (current-slot 0))
+    (while (< current-slot 24)
+      (let* ((current-slot-hour (mod (+ start-hour current-slot) 24))
+            (current-slot-ones (mod current-slot-hour 10))
+            (current-slot-tens (/ (- current-slot-hour current-slot-ones) 10)))
+        (aset hourline (+ 1 (* current-slot 6)) (+ 48 current-slot-tens)) ;; 48 is ascii for '0'
+        (aset hourline (+ 2 (* current-slot 6)) (+ 48 current-slot-ones)))
+      (cl-incf current-slot))
+    hourline))
+
+(defun org-timeline--generate-timeline ()
+  "Generate the timeline string that will represent current agenda view."
+  (let* ((start-offset (* org-timeline-start-hour 60))
+         (current-time (+ (* 60 (string-to-number (format-time-string "%H")))
+                          (string-to-number (format-time-string "%M"))))
+         (current-offset (/ (- current-time start-offset) 10))
+         (slotline (org-timeline--add-elapsed-face
+                    "|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |"
+                    current-offset))
+         (hourline (org-timeline--add-elapsed-face
+                    (org-timeline--generate-hourline)
+                    current-offset))
+         (timeline (concat hourline "\n" slotline))
+         (tasks (org-timeline--list-tasks)))
     (cl-labels ((get-start-pos (current-line beg) (+ 1 (* current-line (1+ (length slotline))) (/ (- beg start-offset) 10)))
                 (get-end-pos (current-line end) (+ 1 (* current-line (1+ (length slotline))) (/ (- end start-offset) 10))))
       (let ((current-line 1))
         (with-temp-buffer
           (insert timeline)
           (-each tasks
             (-lambda ((beg end face))
-              (while (get-text-property (get-start-pos current-line beg) 'org-timeline-occupied)
-                (cl-incf current-line)
-                (when (> (get-start-pos current-line beg) (point-max))
-                  (save-excursion
-                    (goto-char (point-max))
-                    (insert "\n" slotline))))
-              (let ((start-pos (get-start-pos current-line beg))
-                    (end-pos (get-end-pos current-line end)))
-                (put-text-property start-pos end-pos 'font-lock-face face)
-                (put-text-property start-pos end-pos 'org-timeline-occupied t))
-              (setq current-line 1)))
+              (when (>= beg start-offset)
```

Comment:
I can't decide if this should be here or in the `org-timeline--list-tasks` function, but I think I would prefer it there so that it returns a definite list for later processing.